keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,925 @@
+-->
+There are [many
+reasons](http://bit.ly/k8s-multicluster-conversation-starter-doc) why a K8s user
+may want to split their deployments across multiple clusters, but still retain
+mutual dependencies between workloads running in those clusters. Today the
+cluster is a hard boundary, and a service is opaque to a remote K8s consumer
+that would otherwise be able to make use of metadata (e.g. endpoint toplogy) to
+better direct traffic. To support failover or temporarily during migration,
+users may want to consume services spread across clusters, but today that
+requires non-trivial bespoke solutions.
+
+The Multi-Cluster Services API aims to fix these problems.
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+- Define a minimal API to support service discovery and consumption across clusters.
+  - Consume a service in another cluster.
+  - Consume a service deployed in multiple clusters as a single service.
+- When a service is consumed from another cluster its behavior should be
+  predictable and consistent with how it would be consumed within its own
+  cluster.
+- Create building blocks for multi-cluster tooling.
+- Support multiple implementations.
+- Leave room for future extension and new use cases.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Define specific implementation details beyond general API behavior.
+- Change behavior of single cluster services in any way.
+- Define what NetworkPolicy means for multi-cluster services.
+- Solve mechanics of multi-cluster service orchestration.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+#### Terminology
+
+- **supercluster** - A placeholder name for a group of clusters with a high
+  degree of mutual trust and shared ownership that share services amongst
+  themselves. Membership in a supercluster is symmetric and transitive. The set
+  of member clusters are mutually aware, and agree about their collective
+  association.
+- **mcsd-controller** - A controller that syncs services across clusters and
+  makes them available for multi-cluster service discovery (MCSD) and
+  connectivity. There may be multiple implementations, this doc describes
+  expected common behavior.
+
+We propose a new CRD called `ServiceExport`, used to specify which services
+should be exposed across all clusters in the supercluster. `ServiceExports` must
+be created in each cluster that the underlying `Service` resides in. Creation of
+a `ServiceExport` in a cluster will signify that the `Service` with the same
```

Comment:
My understanding as that there will be `ServiceExport` object in the same namespace as original service, correct? If yes, is it a requirement that the namespace should exist in every cluster? If no, does that mean we create these objects in some supercluster system namespace but provide means to access them by namespaced name?

Reply:
The `ServiceExport` must exist in the same namespace as the original service to export that service, but is not required otherwise. If say cluster A exports my-svc and cluster B does not, the `ServiceExport` will only be in cluster A.

On the import side, we do expect a `ServiceImport` to be created with the same namespaced name in each importing cluster. This KEP assumes that's each cluster in the supercluster. The important thing is that this proposal assumes [namespace sameness](https://github.com/kubernetes/community/blob/master/sig-multicluster/namespace-sameness-position-statement.md), so we don't want a situation where a namespace behaves differently in each cluster.

That said (while trying not to open a can of worms), if a given namespace is entirely prohibited from subset of clusters, it seems reasonable that the implementation may then not create a `ServiceImport` in those clusters. In other words, if cluster A exports app1-ns/my-svc and app1-ns exists in cluster B but is prohibited in cluster C, then a `ServiceImport` for `app1-ns/my-svc` will be created in cluster B but not in cluster C and the service would therefore not be discoverable from cluster C.

Reply:
Prohibited suggests some sort of policy or access control to not allow that namespace. What about scenario where`app1-ns` is not already created in Cluster C and isn't prohibited either, maybe oversight from admin/user? Do we fail with appropriate error/log? In other words, is absence of namespace already being present is also considered as prohibited?

Reply:
You're right - we should discuss this more. If `app1-ns` doesn't exist, I'm not sure we want the MCS controller to automatically create it - that seems like an overstep. At the same time, that's distinctly different from a prohibited namespace. Maybe the answer is to just state in an event/condition on the export which clusters were unable to import and why?

Reply:
> Maybe the answer is to just state in an event/condition on the export which clusters were unable to import and why?

Yes, that sounds like a good option. Export status will let users know if import failed on any and the reason, which will allow them to take corrective action.